File: src/data/questions.js
export const QUESTION_BANK = {
    "SWE": [
        {
            "id": "swe_1",
            "question": "Tell me about a time when you had to debug a critical production issue under time pressure.",
            "tags": ["problem-solving", "pressure", "technical"]
        },
        {
            "id": "swe_2",
            "question": "Describe a situation where you had to refactor legacy code. What was your approach?",
            "tags": ["technical", "initiative", "quality"]
        },
        {
            "id": "swe_3",
            "question": "Tell me about a time when you disagreed with a technical decision made by your team.",
            "tags": ["collaboration", "communication", "conflict"]
        },
        {
            "id": "swe_4",
            "question": "Share an example of when you had to learn a new technology or framework quickly to meet a deadline.",
            "tags": ["learning", "pressure", "adaptability"]
        },
        {
            "id": "swe_5",
            "question": "Describe a project where you had to balance technical excellence with business constraints.",
            "tags": ["decision-making", "technical", "business"]
        },
        {
            "id": "swe_6",
            "question": "Tell me about a time when you received critical feedback on your code. How did you respond?",
            "tags": ["growth", "communication", "adaptability"]
        },
        {
            "id": "swe_7",
            "question": "Describe a situation where you identified and prevented a potential security vulnerability.",
            "tags": ["initiative", "technical", "quality"]
        },
        {
            "id": "swe_8",
            "question": "Tell me about your biggest technical failure and what you learned from it.",
            "tags": ["failure", "learning", "growth"]
        },
        {
            "id": "swe_9",
            "question": "Share an example of when you mentored or helped a junior developer solve a complex problem.",
            "tags": ["leadership", "communication", "collaboration"]
        },
        {
            "id": "swe_10",
            "question": "Describe a time when you had to make a trade-off between code quality and delivery speed.",
            "tags": ["decision-making", "pressure", "quality"]
        },
        {
            "id": "swe_11",
            "question": "Tell me about a time when you improved the performance of a system or application significantly.",
            "tags": ["technical", "initiative", "impact"]
        },
        {
            "id": "swe_12",
            "question": "Describe a situation where you had to advocate for a technical approach that others initially resisted.",
            "tags": ["leadership", "communication", "technical"]
        }
    ],
    "Data Analyst": [
        {
            "id": "da_1",
            "question": "Tell me about a time when your data analysis led to a significant business decision.",
            "tags": ["impact", "business", "technical"]
        },
        {
            "id": "da_2",
            "question": "Describe a situation where you found errors or inconsistencies in data. How did you handle it?",
            "tags": ["problem-solving", "quality", "initiative"]
        },
        {
            "id": "da_3",
            "question": "Tell me about a time when you had to explain complex data insights to non-technical stakeholders.",
            "tags": ["communication", "collaboration", "business"]
        },
        {
            "id": "da_4",
            "question": "Share an example of when you had to work with incomplete or messy data.",
            "tags": ["adaptability", "problem-solving", "technical"]
        },
        {
            "id": "da_5",
            "question": "Describe a project where you had to learn a new analytical tool or technique quickly.",
            "tags": ["learning", "adaptability", "technical"]
        },
        {
            "id": "da_6",
            "question": "Tell me about a time when your initial analysis turned out to be wrong. What did you do?",
            "tags": ["failure", "learning", "integrity"]
        },
        {
            "id": "da_7",
            "question": "Describe a situation where you had to prioritize multiple data requests from different stakeholders.",
            "tags": ["decision-making", "pressure", "collaboration"]
        },
        {
            "id": "da_8",
            "question": "Tell me about a time when you identified a trend or pattern that others had missed.",
            "tags": ["initiative", "impact", "technical"]
        },
        {
            "id": "da_9",
            "question": "Share an example of when you had to challenge a business assumption based on your data analysis.",
            "tags": ["communication", "confidence", "impact"]
        },
        {
            "id": "da_10",
            "question": "Describe a time when you automated a reporting process. What was the impact?",
            "tags": ["initiative", "technical", "efficiency"]
        },
        {
            "id": "da_11",
            "question": "Tell me about a situation where you had to balance analytical depth with tight deadlines.",
            "tags": ["pressure", "decision-making", "quality"]
        },
        {
            "id": "da_12",
            "question": "Describe a time when you collaborated with other teams to solve a data-related problem.",
            "tags": ["collaboration", "communication", "problem-solving"]
        }
    ],
    "Product": [
        {
            "id": "pm_1",
            "question": "Tell me about a time when you had to say no to a feature request from an important stakeholder.",
            "tags": ["decision-making", "communication", "conflict"]
        },
        {
            "id": "pm_2",
            "question": "Describe a product you launched that didn't meet expectations. What did you learn?",
            "tags": ["failure", "learning", "adaptability"]
        },
        {
            "id": "pm_3",
            "question": "Tell me about a time when you had to prioritize features with limited resources.",
            "tags": ["decision-making", "pressure", "strategy"]
        },
        {
            "id": "pm_4",
            "question": "Share an example of when you used data to influence a product decision.",
            "tags": ["impact", "technical", "communication"]
        },
        {
            "id": "pm_5",
            "question": "Describe a situation where you had to balance user needs with business goals.",
            "tags": ["decision-making", "business", "empathy"]
        },
        {
            "id": "pm_6",
            "question": "Tell me about a time when you had to get buy-in from engineering for a challenging project.",
            "tags": ["collaboration", "communication", "leadership"]
        },
        {
            "id": "pm_7",
            "question": "Describe a time when you identified a market opportunity that others had overlooked.",
            "tags": ["initiative", "impact", "strategy"]
        },
        {
            "id": "pm_8",
            "question": "Tell me about a situation where you had to make a product decision with incomplete information.",
            "tags": ["decision-making", "pressure", "risk"]
        },
        {
            "id": "pm_9",
            "question": "Share an example of when you had to pivot your product strategy based on user feedback.",
            "tags": ["adaptability", "learning", "empathy"]
        },
        {
            "id": "pm_10",
            "question": "Describe a time when you had to manage conflicting priorities from different stakeholders.",
            "tags": ["collaboration", "conflict", "communication"]
        },
        {
            "id": "pm_11",
            "question": "Tell me about a product feature you championed that became highly successful.",
            "tags": ["initiative", "impact", "leadership"]
        },
        {
            "id": "pm_12",
            "question": "Describe a situation where you had to work with a difficult team member to deliver a product.",
            "tags": ["collaboration", "conflict", "adaptability"]
        }
    ]
};

export function pickSessionQuestions(role, n = 5) {
    if (!QUESTION_BANK[role]) {
        throw new Error(`Unknown role: ${role}`);
    }

    // Copy array to avoid mutation
    const questions = [...QUESTION_BANK[role]];

    // Simple shuffle
    for (let i = questions.length - 1; i > 0; i--) {
        const j = Math.floor(Math.random() * (i + 1));
        [questions[i], questions[j]] = [questions[j], questions[i]];
    }

    return questions.slice(0, n);
}

export function getAllRoles() {
    return Object.keys(QUESTION_BANK);
}

File: questions.py
import random

QUESTION_BANK = {
    "SWE": [
        {
            "id": "swe-01",
            "question": "Tell me about a time you improved system performance under a tight deadline.",
            "tags": ["failure", "ambiguity"]
        },
        {
            "id": "swe-02",
            "question": "Describe a situation where you debugged a production issue with limited information.",
            "tags": ["failure", "ambiguity"]
        },
        {
            "id": "swe-03",
            "question": "Share a time you disagreed with a technical design and how you handled it.",
            "tags": ["conflict", "teamwork"]
        },
        {
            "id": "swe-04",
            "question": "Tell me about a time you learned a new technology quickly to deliver a feature.",
            "tags": ["ambiguity", "leadership"]
        },
        {
            "id": "swe-05",
            "question": "Describe a project where you balanced code quality with speed.",
            "tags": ["ambiguity", "leadership"]
        },
        {
            "id": "swe-06",
            "question": "Tell me about a time you mentored a teammate or helped them unblock.",
            "tags": ["teamwork", "leadership"]
        },
        {
            "id": "swe-07",
            "question": "Describe a situation where you made a mistake in code and what you did next.",
            "tags": ["failure", "leadership"]
        },
        {
            "id": "swe-08",
            "question": "Share a time you reduced technical debt while still delivering features.",
            "tags": ["leadership", "ambiguity"]
        },
        {
            "id": "swe-09",
            "question": "Tell me about a time you coordinated a cross-team technical change.",
            "tags": ["teamwork", "leadership"]
        },
        {
            "id": "swe-10",
            "question": "Describe a time you handled an on-call incident and communicated status.",
            "tags": ["teamwork", "ambiguity"]
        }
    ],
    "Data Analyst": [
        {
            "id": "da-01",
            "question": "Tell me about a time you turned messy data into a clear recommendation.",
            "tags": ["ambiguity", "leadership"]
        },
        {
            "id": "da-02",
            "question": "Describe a situation where stakeholders disagreed with your analysis.",
            "tags": ["conflict", "teamwork"]
        },
        {
            "id": "da-03",
            "question": "Share a time you automated a reporting process to save time.",
            "tags": ["leadership", "ambiguity"]
        },
        {
            "id": "da-04",
            "question": "Tell me about a time you found an unexpected insight in the data.",
            "tags": ["ambiguity", "leadership"]
        },
        {
            "id": "da-05",
            "question": "Describe a project where you had to define success metrics.",
            "tags": ["ambiguity", "leadership"]
        },
        {
            "id": "da-06",
            "question": "Share a time you balanced speed and accuracy in analysis.",
            "tags": ["ambiguity", "teamwork"]
        },
        {
            "id": "da-07",
            "question": "Tell me about a time you explained complex results to a non-technical audience.",
            "tags": ["teamwork", "leadership"]
        },
        {
            "id": "da-08",
            "question": "Describe a time you had to challenge assumptions with data.",
            "tags": ["conflict", "leadership"]
        },
        {
            "id": "da-09",
            "question": "Tell me about a project where data quality issues caused delays and how you handled it.",
            "tags": ["failure", "ambiguity"]
        },
        {
            "id": "da-10",
            "question": "Share a time you partnered with engineering to improve data pipelines.",
            "tags": ["teamwork", "leadership"]
        }
    ],
    "Product": [
        {
            "id": "pm-01",
            "question": "Tell me about a time you prioritized features with limited resources.",
            "tags": ["ambiguity", "leadership"]
        },
        {
            "id": "pm-02",
            "question": "Describe a situation where you aligned multiple teams around a product goal.",
            "tags": ["teamwork", "leadership"]
        },
        {
            "id": "pm-03",
            "question": "Share a time you handled conflicting stakeholder feedback.",
            "tags": ["conflict", "teamwork"]
        },
        {
            "id": "pm-04",
            "question": "Tell me about a product decision you made with incomplete data.",
            "tags": ["ambiguity", "leadership"]
        },
        {
            "id": "pm-05",
            "question": "Describe a time you improved a product metric through iteration.",
            "tags": ["leadership", "ambiguity"]
        },
        {
            "id": "pm-06",
            "question": "Share a time you learned from a product launch that did not go as planned.",
            "tags": ["failure", "leadership"]
        },
        {
            "id": "pm-07",
            "question": "Tell me about a time you clarified an ambiguous requirement.",
            "tags": ["ambiguity", "teamwork"]
        },
        {
            "id": "pm-08",
            "question": "Describe a time you negotiated scope with engineering or design.",
            "tags": ["conflict", "teamwork"]
        },
        {
            "id": "pm-09",
            "question": "Share a time you used customer research to pivot a roadmap.",
            "tags": ["leadership", "ambiguity"]
        },
        {
            "id": "pm-10",
            "question": "Tell me about a time you owned a difficult trade-off decision.",
            "tags": ["conflict", "leadership"]
        }
    ],
    "General": [
        {
            "id": "gen-01",
            "question": "Tell me about a time you took initiative without being asked.",
            "tags": ["leadership", "ambiguity"]
        },
        {
            "id": "gen-02",
            "question": "Describe a situation where you had to resolve a conflict on your team.",
            "tags": ["conflict", "teamwork"]
        },
        {
            "id": "gen-03",
            "question": "Share a time you failed at something and what you learned from it.",
            "tags": ["failure", "leadership"]
        },
        {
            "id": "gen-04",
            "question": "Tell me about a time you collaborated across teams to hit a goal.",
            "tags": ["teamwork", "leadership"]
        },
        {
            "id": "gen-05",
            "question": "Describe a time you had to deliver results with unclear requirements.",
            "tags": ["ambiguity", "leadership"]
        },
        {
            "id": "gen-06",
            "question": "Share a time you had to persuade others to adopt your idea.",
            "tags": ["conflict", "leadership"]
        },
        {
            "id": "gen-07",
            "question": "Tell me about a time you balanced multiple priorities and deadlines.",
            "tags": ["ambiguity", "teamwork"]
        },
        {
            "id": "gen-08",
            "question": "Describe a time you improved a process or workflow for the team.",
            "tags": ["leadership", "teamwork"]
        },
        {
            "id": "gen-09",
            "question": "Share a time you received tough feedback and how you acted on it.",
            "tags": ["conflict", "leadership"]
        },
        {
            "id": "gen-10",
            "question": "Tell me about a time you made a decision with incomplete information.",
            "tags": ["ambiguity", "leadership"]
        }
    ]
}


def get_roles():
    return sorted(QUESTION_BANK.keys())


def get_questions(role, n=5, seed=None):
    questions = QUESTION_BANK.get(role, [])
    if not questions:
        return []
    rng = random.Random(seed)
    if len(questions) <= n:
        return list(questions)
    return rng.sample(questions, n)
